br3ndonland/inboard
# Refactor configuration files and tests

## Description

The Gunicorn and logging configuration files were technically covered by unit tests, but the individual settings were not tested. This PR will add a dedicated test module for each configuration file, and add tests to verify the Gunicorn and logging settings.

One of the major benefits of unit testing is that the tests not only verify that code is working as expected, but also reveal where the code is _not_ working as expected. This was the case here, and these updates result in cleaner and more correct source code.

## Changes

- Move Gunicorn conf tests to separate module (d28753b)
- **Refactor Gunicorn worker calculation and tests** (315c1c1)
  - The `gunicorn_conf.calculate_workers()` method returns a desired number of Gunicorn workers, based on the arguments passed in. The arguments are read from environment variables.
  - The calculation is deceptively complex, and a variety of edge cases emerged over time. The tests were correspondingly complex and convoluted, testing various interrelated conditions, and becoming less and less readable.
  - This commit will refactor the Gunicorn worker calculation and tests. The result is more readable (and correct) code.
- Organize Gunicorn settings (aaef2f0): The same result is obtained with less than half the lines of code (29 -> 14).
- **Test Gunicorn settings** (324f88f)
  - Start Gunicorn with the [`--print-config` setting](https://docs.gunicorn.org/en/latest/settings.html#print-config) to load and output the configuration.
  - Use the [`capfd` pytest fixture](https://docs.pytest.org/en/latest/how-to/capture-stdout-stderr.html) to read the Gunicorn output.
  - These could be considered integration tests, because Gunicorn is actually starting and loading the settings. It also means they are substantially slower than the other tests, but the entire test suite still runs in only a few seconds.
- Use similar syntax to set up Gunicorn and Uvicorn (1b739ed)
- **Consolidate logging configuration method and tests** (8cd8db2)
  - The `configure_logging()` method was previously located in `start.py`, but the logging configuration dictionary was in `logging_conf.py`. This organization was not ideal for readability and separation of concerns. It also created logistical complications, such as requiring the Gunicorn configuration file `gunicorn_conf.py` to import from start.py, so that it could configure logging for Gunicorn. The `start.py` script should be self-contained, and should import other modules and objects from the package, not the other way around.
  - This commit will move `configure_logging()` to `logging_conf.py`, and move the tests to a corresponding module, `test_logging_conf.py`. This matches up nicely with the latest updates to the `gunicorn_conf.py`, by having a setup method at the top of the module, and the settings below.
- Correct attribute error on module spec loader (0ba94fc)
- **Split up logging configuration method** (373c98d)
  - The `configure_logging()` method did multiple things: find and import a logging configuration module, load the logging configuration dictionary, and apply the dictionary configuration to the Python logger.
  - This commit will refactor `configure_logging()` into two methods: `find_and_load_logging_conf()` will return a dictionary configuration, and `configure_logging()` will apply the dictionary configuration.
  - The return value of `configure_logging()` will remain the same, so there will not be any changes to the programming API.
  - Also, now that the `configure_logging()` method is in the same module as the logging configuration dictionary, that dictionary will be used by default, instead of finding the module separately.
- Remove prefix from mocker objects in logger tests (9a61fc2)
- **Test logger output after configuring logging** (273f795)
  - In addition to testing the methods that load the logging configuration, it can also be useful to test that the logger messages have the expected format. This commit will add tests that output logger messages in various formats and assert that each output matches the expected format.
  - As in the tests of the Gunicorn server output, these tests will use the [`capfd` pytest fixture](https://docs.pytest.org/en/latest/how-to/capture-stdout-stderr.html).
  - It is also possible to use the [`caplog` fixture](https://docs.pytest.org/en/latest/how-to/logging.html), but I have not had as much success with it.
- Update links to pytest docs (aa243dd): Love the new docs!
- Clarify Gunicorn variable descriptions in README (40baca4)

## Related

br3ndonland/inboard#3
br3ndonland/inboard#11

- [x] I have reviewed the [Guidelines for Contributing](https://github.com/br3ndonland/inboard/blob/develop/.github/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/br3ndonland/inboard/blob/develop/.github/CODE_OF_CONDUCT.md).
